### ISSUE TITLE
Result query order is inconsistent for "ID" column across versions

### ISSUE THREAD

**Issue by Orenef11** [scylladb/scylla#8207]
I notice that there is an inconsistency in the output for the `cql_additional_tests.TestCQL.nameless_index_test` test.
For each of the following versions of `4.3` (`4.3.0`, `4.3.1`, `4.3RC0`, `4.3RC1`, and `4.3RC3`) I get the wrong result `['Bob'], ['Tom']
`.
For versions `4.4.rc1-0.20210131.fd868722d` and `unstable/master:2021-02-15T01:57:05Z`(`4.5` version) I get the result `['Tom'], ['Bob']`
The test was created on 02/12/2015 and I do not think we have had this problem until now ...

Also, I do not know why, but I run the following scenario for the `scylladb/scylla:4.3.0` and `scylladb/scylla:4.3.1`, and the result is correct.

```
docker rm -f some-scylla
docker run --name some-scylla --hostname some-scylla -d scylladb/scylla:4.3.0
docker exec -it some-scylla cqlsh
CREATE KEYSPACE ks2 WITH replication = {'class': 'SimpleStrategy', 'replication_factor' : 3};
USE ks2;
CREATE TABLE nameless_index_test (id text PRIMARY KEY, birth_year int);
INSERT INTO nameless_index_test (id, birth_year) VALUES ('Tom', 42);
INSERT INTO nameless_index_test (id, birth_year) VALUES ('Paul', 24);
INSERT INTO nameless_index_test (id, birth_year) VALUES ('Bob', 42);
SELECT id FROM nameless_index_test WHERE birth_year = 42 ALLOW FILTERING;

The result is:
id
-----
 Tom
 Bob
```

**Comment by Orenef11** [comment]
@roydahan @fruch 
I notice that a test failed for version `4.3` and not for version `4.4`.

I ran the `cql_additional_tests.TestCQL.nameless_index_test` test for following versions: `4.3.rc3-0.20201217.5bd52e4db`, `4.4` and `4.5`:
* `4.3.rc3-0.20201217.5bd52e4db` - The test is failed because the order of data is wrong (`[['Bob'], ['Tom']]`)
* `4.4.rc1-0.20210131.fd868722d` - The test is passed
* `4.5.dev-0.20210214.9e029f09e` - The test is failed.

Also, I tested the same scenario for the latest versions of `4.3.1-0.20210210.46650adcd` and `4.4.rc1-0.20210131.fd868722d` and passed a test.
```
docker rm -f some-scylla
docker run --name some-scylla --hostname some-scylla -d scylladb/scylla:4.3.1
docker exec -it some-scylla cqlsh
CREATE KEYSPACE ks2 WITH replication = {'class': 'SimpleStrategy', 'replication_factor' : 3};
USE ks2;
CREATE TABLE nameless_index_test (id text PRIMARY KEY, birth_year int);
INSERT INTO nameless_index_test (id, birth_year) VALUES ('Tom', 42);
INSERT INTO nameless_index_test (id, birth_year) VALUES ('Paul', 24);
INSERT INTO nameless_index_test (id, birth_year) VALUES ('Bob', 42);
SELECT id FROM nameless_index_test WHERE birth_year = 42 ALLOW FILTERING;

The result is:
id
-----
 Tom
 Bob


docker rm -f some-scylla
docker run --name some-scylla --hostname some-scylla -d scylladb/scylla:4.4.rc1
docker exec -it some-scylla cqlsh
CREATE KEYSPACE ks1 WITH replication = {'class': 'SimpleStrategy', 'replication_factor' : 1};
USE ks1;
CREATE TABLE nameless_index_test (id text PRIMARY KEY, birth_year int);
INSERT INTO nameless_index_test (id, birth_year) VALUES ('Tom', 42);
INSERT INTO nameless_index_test (id, birth_year) VALUES ('Paul', 24);
INSERT INTO nameless_index_test (id, birth_year) VALUES ('Bob', 42);
SELECT id FROM nameless_index_test WHERE birth_year = 42 ALLOW FILTERING;

The result is:
id
-----
 Tom
 Bob
```

Therefore, the change made in SCT caused this problem.

**Comment by roydahan** [comment]
@juliayakovlev It looks like you change in SCT the order of the results.
The actual question is why 4.4 behaves differently than previous versions in terms of the order?
Which Scylla commit changes that and which one is the correct order? (AND If the order actually matters).

**Comment by Orenef11** [comment]
> The actual question is why 4.4 behaves differently than previous versions in terms of the order?
> Which Scylla commit changes that and which one is the correct order? (AND If the order actually matters).

The last three versions `4.3`, `4.4` and `4.5` (master) behave as required.
I do not know what happened in the versions before that, but in the tests, I did with the docker image, I got the correct order (`[["Tom"], ["Bob"]]`).

Just for the `4.3.1-0.20210210.46650adcd` version, I got an incorrect result ((`[["Bob"], ["Tom"]]`)).
But, since version 4.3.1 fixed the problem, is it necessary to understand what happened before this version?

**Comment by roydahan** [comment]
AFAIU from Oren he notice that even between 4.3.0 to 4.3.1 there is a difference, so probably there is a commit that was backport to change it.

mom closing the issue and we will find the commit just to understand it.

**Comment by nyh** [comment]
Interesting issue. What you are doing here is a whole-table scan and then expecting the order of **partitions** to be a specific order. In theory, you're not supposed to test that - Scylla users are not supposed to know or care about partition order, which is determined through some hash function (the so-called "partitioner") which they don't control. So either order is technically correct.

BUT, the order needs to be fixed, in the sense that you can't have different nodes disagreeing on the order - or the result will a big mess when trying to reconcile scans on different replicas. A corollary of not having different nodes disagreeing on the partition order is that we can't change the partition order between Scylla versions (well, we can, with a "feature bit", but it's complicated). So we don't want this order to change between version - and if it did, it sounds bad. 

But that being said, we haven't, as far as I know, changed the partitioner function in years, so I wouldn't have expected the order of the results to have changed in 4.3.0 and then changed back in 4.3.1. We did change some internals of the partitioner implementation in the last year, so maybe @avikivity or @haaawk remember if we had a short-lived regression in that area that affected 4.3.0 but then got fixed in 4.3.1?

**Comment by Orenef11** [comment]
> Interesting issue. What you are doing here is a whole-table scan and then expecting the order of **partitions** to be a specific order. In theory, you're not supposed to test that - Scylla users are not supposed to know or care about partition order, which is determined through some hash function (the so-called "partitioner") which they don't control. So either order is technically correct.
> 
> BUT, the order needs to be fixed, in the sense that you can't have different nodes disagreeing on the order - or the result will a big mess when trying to reconcile scans on different replicas. A corollary of not having different nodes disagreeing on the partition order is that we can't change the partition order between Scylla versions (well, we can, with a "feature bit", but it's complicated). So we don't want this order to change between version - and if it did, it sounds bad.
> 
> But that being said, we haven't, as far as I know, changed the partitioner function in years, so I wouldn't have expected the order of the results to have changed in 4.3.0 and then changed back in 4.3.1. We did change some internals of the partitioner implementation in the last year, so maybe @avikivity or @haaawk remember if we had a short-lived regression in that area that affected 4.3.0 but then got fixed in 4.3.1?

Sorry for the confusion, I corrected the problem description.
The test in the `dtest` fails for each of the versions: `4.3.0`, `4.3.1`, `4.3RC0`, `4.3RC1`, and `4.3RC3` (I run the test on top of Docker).

But, I do the following scenario, it returns a correct result:
```
docker rm -f some-scylla
docker run --name some-scylla --hostname some-scylla -d scylladb/scylla:4.3.0
docker exec -it some-scylla cqlsh
CREATE KEYSPACE ks2 WITH replication = {'class': 'SimpleStrategy', 'replication_factor' : 3};
USE ks2;
CREATE TABLE nameless_index_test (id text PRIMARY KEY, birth_year int);
INSERT INTO nameless_index_test (id, birth_year) VALUES ('Tom', 42);
INSERT INTO nameless_index_test (id, birth_year) VALUES ('Paul', 24);
INSERT INTO nameless_index_test (id, birth_year) VALUES ('Bob', 42);
SELECT id FROM nameless_index_test WHERE birth_year = 42 ALLOW FILTERING;

The result is:
id
-----
 Tom
 Bob
```
Thus, I also got confused and thought that for `4.3.1` it works.
I will try to check what actions we are doing in the dtest that I did not do in the scenario I wrote for Docker.

@fruch @nyh 
Is there a way to get all the commands we executed during the test?

**Comment by dekimir** [comment]
The dtest is a single-node test, and as such it is wrong to expect any particular order without using ORDER BY in the query.

**Comment by nyh** [comment]
> The dtest is a single-node test, and as such it is wrong to expect any particular order without using ORDER BY in the query.

Although you shouldn't care about the order, I'm worried that this order changed between 4.3 and 4.4. Doesn't it mean that if we do have multiple nodes, and a mixed cluster (some 4.3 nodes, some 4.4 nodes), we can get broken reconciliation of scans?

I'll try to find time to test this myself later with cql-pytest (as you said, it's a single-node test).  
If this is a real thing, and not some false-alarm, I think we need to understand from the partitioning experts (@avikivity ? @denesb ? @haaawk ?) if anything changed in the partitioning function between 4.3 and 4.4 - and if it did, how are upgrades supposed to work (it's even worse than mixed clusters - the entire sstable is wrong if the partitioning order changes).

@orenf11, why is this issue closed?

**Comment by Orenef11** [comment]
> > The dtest is a single-node test, and as such it is wrong to expect any particular order without using ORDER BY in the query.
> 
> Although you shouldn't care about the order, I'm worried that this order changed between 4.3 and 4.4. Doesn't it mean that if we do have multiple nodes, and a mixed cluster (some 4.3 nodes, some 4.4 nodes), we can get broken reconciliation of scans?
> 
> I'll try to find time to test this myself later with cql-pytest (as you said, it's a single-node test).
> If this is a real thing, and not some false-alarm, I think we need to understand from the partitioning experts (@avikivity ? @denesb ? @haaawk ?) if anything changed in the partitioning function between 4.3 and 4.4 - and if it did, how are upgrades supposed to work (it's even worse than mixed clusters - the entire sstable is wrong if the partitioning order changes).
> 
> @orenf11, why is this issue closed?

I reopen the bug.

**Comment by nyh** [comment]
@Orenef11 so I'm confused now. I don't understand how you reproduced this bug or on which version...

I just wrote a test that creates partitions and SELECTS all of them, and unsurprisingly (for reasons I explain above), Scylla 4.3.0, Scylla master and Cassandra all return exactly the same order. So it **looks** like this bug doesn't exist - so perhaps this issue should be closed yet again. If the bug does exist, can you please explain how to reproduce it?

The cql-pytest I just which passes (produces the same order) on Scylla 4.3.0, Scylla master and Cassandra trunk:
```python
# In theory, partitions are ordered in some unknown order and can change.
# But it cannot really change: The order must be fixed for a partitioner with
# a given name, because if the partitioning function changes, old sstables
# can no longer be read because they have the "wrong" partition order.
# So we can be sure that the meaning of a specific 
def test_filter_on_unset(cql, test_keyspace):
    with new_test_table(cql, test_keyspace, "k int PRIMARY KEY") as table:
        stmt = cql.prepare(f"INSERT INTO {table} (k) VALUES (?)")
        for i in range(10):
            cql.execute(stmt, [i])
        assert list(cql.execute(f"SELECT k FROM {table}")) == [(5,), (1,), (8,), (0,), (2,), (4,), (7,), (6,), (9,), (3,)]
```

**Comment by Orenef11** [comment]
@nyh 
Like I said before, I do not know what the test is written in dtest does, but to reproduce this issue you need to run `cql_additional_tests.TestCQL.nameless_index_test` test.

I still could not understand why the problem only for `cql_additional_tests.TestCQL.nameless_index_test` test., and why the following test is passed:
```
docker rm -f some-scylla
docker run --name some-scylla --hostname some-scylla -d scylladb/scylla:4.3.0
docker exec -it some-scylla cqlsh
CREATE KEYSPACE ks2 WITH replication = {'class': 'SimpleStrategy', 'replication_factor' : 3};
USE ks2;
CREATE TABLE nameless_index_test (id text PRIMARY KEY, birth_year int);
INSERT INTO nameless_index_test (id, birth_year) VALUES ('Tom', 42);
INSERT INTO nameless_index_test (id, birth_year) VALUES ('Paul', 24);
INSERT INTO nameless_index_test (id, birth_year) VALUES ('Bob', 42);
SELECT id FROM nameless_index_test WHERE birth_year = 42 ALLOW FILTERING;

The result is:
id
-----
 Tom
 Bob
```

**Comment by denesb** [comment]
@Orenef11 a good way to verify whether there is a real issue or not is to also select the respective token of the partition key:
```
SELECT token(id), id FROM nameless_index_test WHERE birth_year = 42 ALLOW FILTERING;
```
Partitions are ordered by the token of their partition key. This experiment will help us answer a few questions: are partitions indeed out-of-order w.r.t. their tokens? If not, have their tokens changed between the versions?

**Comment by nyh** [comment]
I don't understand what the problem is so I'll try to reproduce it myself (@Orenef11 explained to me which dtest I need to run) and see if I can come up with any conclusions. @eliransin FYI.

**Comment by Orenef11** [comment]
> @Orenef11 a good way to verify whether there is a real issue or not is to also select the respective token of the partition key:
> 
> ```
> SELECT token(id), id FROM nameless_index_test WHERE birth_year = 42 ALLOW FILTERING;
> ```
> 
> Partitions are ordered by the token of their partition key. This experiment will help us answer a few questions: are partitions indeed out-of-order w.r.t. their tokens? If not, have their tokens changed between the versions?

I ran the new query for the following versions:
**4.3.2-0.20210301.5cdc1fa66 with build-id 3a7f9a6b65ec73bba0a38f3f349a91c49129c598**:  `[[4645660266327417866, 'Bob'], [-5746897751862998708, 'Tom']]`

**4.4.rc1-0.20210131.fd868722d with build-id 81a1e1af8194f89648f0de52d16221ac170c539a**: `[[-5746897751862998708, 'Tom'], [4645660266327417866, 'Bob']]`

**4.5.dev-0.20210214.9e029f09e with build-id 25a27807328f19ca39976d15d3780b63e6c25a97**: `[[-5746897751862998708, 'Tom'], [4645660266327417866, 'Bob']]`

**Comment by nyh** [comment]
@Orenef11 wow, so the tokens in 4.3.2 are correct, but they are sorted wrong! It is possible this isn't a general "sorting wrong" case but specific to positive vs. negative tokens. What I don't understand is how we didn't see disastrous results if this was the case - wouldn't upgrading from 4.2 to 4.3 - or from 4.3 to 4.4 - where the tokens are sorted differently - create huge problems for all sorts of algorithms which assume the sstables are, as their name suggests, sorted? 

I can confirm your previous observation - that dtest `cql_additional_tests.py:TestCQL.nameless_index_test` fails on 4.3.1 but passes on master. I'll now try to confirm and test your new more specific observation. This is very helpful, thanks.

**Comment by nyh** [comment]
The good news is that this is not wrong tokens (as @Orenef11 already proved) but also not wrong-token-sorting-everywhere as I feared, because the following simple cql-pytest test, which tests a full table scan and checks the order of the results, passes on master, Cassandra, **and** on 4.3.1:
```python
def test_scan_order_str(cql, test_keyspace):
    with new_test_table(cql, test_keyspace, "s text PRIMARY KEY") as table:
        stmt = cql.prepare(f"INSERT INTO {table} (s) VALUES (?)")
        for i in ['Tom', 'Bob', 'Donald', 'Mickey']:
            cql.execute(stmt, [i])
        assert list(cql.execute(f"SELECT TOKEN(s), s FROM {table}")) == [
            (-8345977921753544684, 'Donald'),
            (-5746897751862998708, 'Tom'),
            ( 1592152328590759703, 'Mickey'),
            ( 4645660266327417866, 'Bob'),
        ]
```

So if the sorting here is correct, but in the dtest it isn't, what's the difference? Filtering? I'll continue to investigate.

**Comment by nyh** [comment]
OH!
I finally realized (I should have guessed by the name!!!) that we are scanning here through an index.

This issue is a duplicate of https://github.com/scylladb/scylla/issues/7443
It was fixed in 4.4, and @avikivity decided not to backport to 4.3, because the only bad effect is the partition order.


Believe it or not, I even wrote a cql-pytest for this bug, `test_secondary_index.py::test_partition_order_with_si`, and it does fail on 4.3.

Two things I didn't realize when we didn't backport this fix:
1. That it was a regression from 4.2 that did this correctly (is it? I didn't test now but Oren says the dtest didn't fail on 4.2 IIRC)
2. I didn't think how a mixed cluster, with part 4.3 and part 4.4, can handle such a request, with each replica returning a different order.

**Comment by Orenef11** [comment]
> OH!
> I finally realized (I should have guessed by the name!!!) that we are scanning here through an index.
> 
> This issue is a duplicate of #7443
> It was fixed in 4.4, and @avikivity decided not to backport to 4.3, because the only bad effect is the partition order.
> 
> Believe it or not, I even wrote a cql-pytest for this bug, `test_secondary_index.py::test_partition_order_with_si`, and it does fail on 4.3.
> 
> Two things I didn't realize when we didn't backport this fix:
> 
> 1. That it was a regression from 4.2 that did this correctly (is it? I didn't test now but Oren says the dtest didn't fail on 4.2 IIRC)
> 2. I didn't think how a mixed cluster, with part 4.3 and part 4.4, can handle such a request, with each replica returning a different order.

1. for `4.2` we got the same result as `4.3`: `[[4645660266327417866, 'Bob'], [-5746897751862998708, 'Tom']]`
2. I do not know what the result will be, because to date we have failed before the upgrade.
There was one time when we did a `rollback` to the node (after two successful upgrades), but I'm not sure it's because of this error.

**Comment by Orenef11** [comment]
I closed the issue because it is a duplicate of https://github.com/scylladb/scylla/issues/7443

**Comment by nyh** [comment]
Oren, yes, thanks, I misunderstood about 4.2 So it wasn't a regression. About my second observation, the fix for issue #7443, commit 7ff72b0ba575766224848749fffc892eef166e78, paid special attention to rolling upgrades. An index created in 4.2 and then upgraded to 4.3 will retain its old "wrong" order forever (or at least should, I don't know how carefully we tested this part of the implementation), so there is no problem in making a request which comes from two replicas with two versions.
So it's not critical that we backport this fix to 4.3. We only need to backport it if we think that it's important that 4.3 users do not experience issue #7443.
